File: generation/py_src/classes.py
class particle:
  __slots__ = 'x', 'y', 'z', 'dx', 'dy', 'dz'
  
  def __init__(self, *args):
    if len(args) == 6:
      self.x = args[0]; self.y = args[1]; self.z = args[2]
      self.dx = args[3]; self.dy = args[4]; self.dz = args[5]
    elif len(args) == 1:
      self.x = args[0].x; self.y = args[0].y; self.z = args[0].z
      self.dx = args[0].dx; self.dy = args[0].dy; self.dz = args[0].dz
  
  def distance_to_squared(self, other):
    return (other.x - self.x) ** 2 + (other.y - self.y) ** 2 + (other.z - self.z) ** 2
  
  def distance_to(self, other):
    return self.distance_to_squared(other) ** 0.5
  
  def vector_away_from_other(self, other, length):
    # multiply difference between the 2 particles' positions, by 1 / distance, then by length
    
    # calculate this multiplication constant
    distance = self.distance_to(other)
    multiplication_const = 1 / distance * length if distance != 0 else 0
    
    return (
      (self.x - other.x) * multiplication_const,
      (self.y - other.y) * multiplication_const,
      (self.z - other.z) * multiplication_const
    )
  
  def apply_velocity(self, dx, dy, dz, time_step):
    return particle(
      self.x + dx * time_step,
      self.y + dy * time_step,
      self.z + dz * time_step,
      self.dx,
      self.dy,
      self.dz
    )
  
  def apply_own_velocity(self, time_step):
    return self.apply_velocity(self.dx, self.dy, self.dz, time_step)
  
  def apply_acceleration(self, ddx, ddy, ddz, time_step):
    return particle(
      self.x,
      self.y,
      self.z,
      self.dx + ddx * time_step,
      self.dy + ddy * time_step,
      self.dz + ddz * time_step,
    )

class system_state:
  __slots__ = 'time', 'particles'
  
  def __init__(self, time, particles):
    self.time = time
    self.particles = particles

class simulation_params:
  __slots__ = (
    'particle_radius',
    'particle_mass',
    'grav_constant',
    'lennard_jones_well_depth',
    'linear_damping_multiplier',
    'time_step',
    'num_steps',
    'particle_configuration',
    'csv_file_skip_steps',
    'status_update_skip_steps',
  )
  
  def __init__(
    self,
    particle_radius,
    particle_mass,
    grav_constant,
    lennard_jones_well_depth,
    linear_damping_multiplier,
    time_step,
    num_steps,
    particle_configuration,
    csv_file_skip_steps,
    status_update_skip_steps
  ):
    self.particle_radius = particle_radius
    self.particle_mass = particle_mass
    self.grav_constant = grav_constant
    self.lennard_jones_well_depth = lennard_jones_well_depth
    self.linear_damping_multiplier = linear_damping_multiplier
    self.time_step = time_step
    self.num_steps = num_steps
    self.particle_configuration = particle_configuration
    self.csv_file_skip_steps = csv_file_skip_steps
    self.status_update_skip_steps = status_update_skip_steps

File: generation/rust_src/src/classes.rs
#[derive(Clone)]
pub struct Particle {
  pub x: f64,
  pub y: f64,
  pub z: f64,
  pub dx: f64,
  pub dy: f64,
  pub dz: f64,
}

impl Particle {
  pub fn from_coords(
    x: f64, y: f64, z: f64,
    dx: f64, dy: f64, dz: f64
  ) -> Particle {
    Particle {
      x,
      y,
      z,
      dx,
      dy,
      dz,
    }
  }
  
  pub fn from_particle(particle_obj: &Particle) -> Particle {
    Particle {
      x: particle_obj.x,
      y: particle_obj.y,
      z: particle_obj.z,
      dx: particle_obj.dx,
      dy: particle_obj.dy,
      dz: particle_obj.dz,
    }
  }
  
  pub fn distance_to_squared(&self, other: &Particle) -> f64 {
    let delta_x = other.x - self.x;
    let delta_y = other.y - self.y;
    let delta_z = other.z - self.z;
    
    delta_x * delta_x + delta_y * delta_y + delta_z * delta_z
  }
  
  pub fn distance_to(&self, other: &Particle) -> f64 {
    f64::sqrt(self.distance_to_squared(other))
  }
  
  pub fn vector_away_from_other(&self, other: &Particle, length: f64) -> (f64, f64, f64) {
    // multiply difference between the 2 particles' positions, by 1 / distance, then by length
    
    // calculate this multiplication constant
    let distance = self.distance_to(other);
    let multiplication_const = if distance != 0.0 {
      1.0 / distance * length
    } else {
      0.0
    };
    
    (
      (self.x - other.x) * multiplication_const,
      (self.y - other.y) * multiplication_const,
      (self.z - other.z) * multiplication_const
    )
  }
  
  pub fn apply_velocity(
    &self,
    dx: f64, dy: f64, dz: f64,
    time_step: f64,
  ) -> Particle {
    Particle {
      x: self.x + dx * time_step,
      y: self.y + dy * time_step,
      z: self.z + dz * time_step,
      dx: self.dx,
      dy: self.dy,
      dz: self.dz,
    }
  }
  
  pub fn apply_own_velocity(&self, time_step: f64) -> Particle {
    self.apply_velocity(self.dx, self.dy, self.dz, time_step)
  }
  
  pub fn apply_acceleration(&self, ddx: f64, ddy: f64, ddz: f64, time_step: f64) -> Particle {
    Particle {
      x: self.x,
      y: self.y,
      z: self.z,
      dx: self.dx + ddx * time_step,
      dy: self.dy + ddy * time_step,
      dz: self.dz + ddz * time_step,
    }
  }
}

pub struct SystemState {
  pub time: f64,
  pub particles: Vec<Particle>,
}

pub struct SimulationParams {
  pub particle_radius: f64,
  pub particle_mass: f64,
  pub grav_constant: f64,
  pub lennard_jones_well_depth: f64,
  pub linear_damping_multiplier: f64,
  pub time_step: f64,
  pub num_steps: u64,
  pub particle_configuration: u8,
  pub csv_file_skip_steps: u64,
  pub status_update_skip_steps: u64,
}
